snorkel/contrib/fonduer/fonduer/lf_helpers.py

```diff
@@ -0,0 +1,1045 @@
+                 if abs(phrase.phrase_num - span.sentence.phrase_num) <= d and phrase != span.sentence]):
+            yield ngram
+
+
+def get_cell_ngrams(span, attrib='words', n_min=1, n_max=1, lower=True):
+    """Get the ngrams that are in the Cell of the given span, not including itself.
+
+    Note that if a candidate is passed in, all of its Spans will be searched.
+
+    :param span: The span whose Cell is being searched
+    :param attrib: The token attribute type (e.g. words, lemmas, poses)
+    :param n_min: The minimum n of the ngrams that should be returned
+    :param n_max: The maximum n of the ngrams that should be returned
+    :param lower: If True, all ngrams will be returned in lower case
+    :rtype: a _generator_ of ngrams
+    """
+    spans = [span] if isinstance(span, TemporarySpan) else span.get_contexts()
+    for span in spans:
+        for ngram in get_phrase_ngrams(span, attrib=attrib, n_min=n_min, n_max=n_max, lower=lower):
+            yield ngram
+        if isinstance(span.sentence, Phrase) and span.sentence.cell is not None:
+            for ngram in chain.from_iterable(
+                    [tokens_to_ngrams(getattr(phrase, attrib), n_min=n_min, n_max=n_max, lower=lower)
+                     for phrase in span.sentence.cell.phrases if phrase != span.sentence]):
+                yield ngram
+
+
+def get_neighbor_cell_ngrams(span, dist=1, directions=False, attrib='words', n_min=1, n_max=1, lower=True):
+    """Get the ngrams from all Cells that are within a given Cell distance in one direction from the given Span.
+
+    Note that if a candidate is passed in, all of its Spans will be searched.
+    If `directions=True``, each ngram will be returned with a direction in {'UP', 'DOWN', 'LEFT', 'RIGHT'}.
+
+    :param span: The span whose neighbor Cells are being searched
+    :param dist: The Cell distance within which a neighbor Cell must be to be considered
+    :param directions: A Boolean expressing whether or not to return the direction of each ngram
+    :param attrib: The token attribute type (e.g. words, lemmas, poses)
+    :param n_min: The minimum n of the ngrams that should be returned
+    :param n_max: The maximum n of the ngrams that should be returned
+    :param lower: If True, all ngrams will be returned in lower case
+    :rtype: a _generator_ of ngrams (or (ngram, direction) tuples if directions=True)
+    """
+    # TODO: Fix this to be more efficient (optimize with SQL query)
+    spans = [span] if isinstance(span, TemporarySpan) else span.get_contexts()
+    for span in spans:
+        for ngram in get_phrase_ngrams(span, attrib=attrib, n_min=n_min, n_max=n_max, lower=lower):
+            yield ngram
+        if isinstance(span.sentence, Phrase) and span.sentence.cell is not None:
+            root_cell = span.sentence.cell
+            for phrase in chain.from_iterable([_get_aligned_phrases(phrase, 'row'), _get_aligned_phrases(phrase, 'col')]):
```

Comment:
`phrase` is undefined and this line will throw an error when called. We will need to fix this.

```
---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
<ipython-input-46-e9231ec49e5f> in <module>()
----> 1 LF_test(test_cands[0])

<ipython-input-45-35a0ac223681> in LF_test(c)
      1 def LF_test(c):
----> 2     neighbor_ngrams = list(get_neighbor_cell_ngrams(c.attr))
      3     print neighbor_ngrams

/home/lwhsiao/repos/snorkel/snorkel/contrib/fonduer/fonduer/lf_helpers.py in get_neighbor_cell_ngrams(c, dist, directions, attrib, n_min, n_max, lower)
    291         if isinstance(span.sentence, Phrase) and span.sentence.cell is not None:
    292             root_cell = span.sentence.cell
--> 293             for phrase in chain.from_iterable([_get_aligned_phrases(phrase, 'row'), _get_aligned_phrases(phrase, 'col')]):
    294                 row_diff = min_row_diff(phrase, root_cell, absolute=False)
    295                 col_diff = min_col_diff(phrase, root_cell, absolute=False)

UnboundLocalError: local variable 'phrase' referenced before assignment
```